packages/block-library/src/paragraph/drop-zone.js

```diff
@@ -0,0 +1,90 @@
+/**
+ * WordPress dependencies
+ */
+import { useState } from '@wordpress/element';
+import { useSelect } from '@wordpress/data';
+import {
+	__experimentalUseOnBlockDrop as useOnBlockDrop,
+	store as blockEditorStore,
+} from '@wordpress/block-editor';
+import {
+	__experimentalUseDropZone as useDropZone,
+	useReducedMotion,
+} from '@wordpress/compose';
+import {
+	Popover,
+	__unstableMotion as motion,
+	__unstableAnimatePresence as AnimatePresence,
+} from '@wordpress/components';
+
+const animateVariants = {
+	hide: { opacity: 0, scaleY: 0.75 },
+	show: { opacity: 1, scaleY: 1 },
+	exit: { opacity: 0, scaleY: 0.9 },
+};
+
+export default function DropZone( { paragraphElement, clientId } ) {
+	const { rootClientId, blockIndex } = useSelect(
+		( select ) => {
+			const selectors = select( blockEditorStore );
+			return {
+				rootClientId: selectors.getBlockRootClientId( clientId ),
+				blockIndex: selectors.getBlockIndex( clientId ),
+			};
+		},
+		[ clientId ]
+	);
+	const onBlockDrop = useOnBlockDrop( rootClientId, blockIndex, true );
+	const [ isVisible, setIsVisible ] = useState( false );
+	const dropZoneRef = useDropZone( {
+		onDrop: onBlockDrop,
+		onDragEnter: () => {
+			setIsVisible( true );
+		},
+		onDragLeave: () => {
+			setIsVisible( false );
+		},
+	} );
+	const reducedMotion = useReducedMotion();
+
+	return (
+		<Popover
+			anchor={ paragraphElement }
+			animate={ false }
+			placement="top-start"
+			focusOnMount={ false }
+			flip={ false }
+			resize={ false }
+			className="wp-block-paragraph__drop-zone"
+		>
+			<div
+				ref={ dropZoneRef }
+				className="wp-block-paragraph__drop-zone-backdrop"
+				style={ {
+					width: paragraphElement?.offsetWidth,
+					height: paragraphElement?.offsetHeight,
+				} }
+			>
+				<AnimatePresence>
+					{ isVisible ? (
+						<motion.div
+							key="drop-zone-foreground"
+							initial={
+								reducedMotion
+									? animateVariants.show
+									: animateVariants.hide
+							}
+							animate={ animateVariants.show }
+							exit={
+								reducedMotion
+									? animateVariants.show
+									: animateVariants.exit
+							}
+							className="wp-block-paragraph__drop-zone-foreground"
+						/>
+					) : null }
+				</AnimatePresence>
+			</div>
+		</Popover>
```

Comment:
It's a shame that so much of this has to be implemented as a seperate component to the components' package `DropZone`, but I see why.

Do you think there's any opportunity to consolidate the two in the future? The thing with the components package version is that it's really legacy code as it implicitly calls the `useDropZone` hook (it's like that for purely back compat reasons).

I think ideally there would be some purely presentational version of the `DropZone` component.